File: src/models/corpse.ts
class Corpse {
  name: string;
  qualifier: string | undefined;
  birthDate: Date | undefined;
  deathDate: Date;
  description: string;
  link: string;

  public constructor(
    name: string,
    qualifier: string | undefined,
    birthDate: Date | undefined,
    deathDate: Date,
    description: string,
    link: string,
  ) {
    this.name = name;
    this.qualifier = qualifier;
    this.birthDate = birthDate && new Date(birthDate);
    this.deathDate = new Date(deathDate);
    this.description = description;
    this.link = link;
  }
}

const getAge = (start: Date, end: Date): { age: number; period: string } => {
  let years = end.getFullYear() - start.getFullYear();
  if (
    end.getMonth() < start.getMonth() ||
    (end.getMonth() === start.getMonth() && end.getDate() < start.getDate())
  ) {
    years--;
  }

  if (years >= 1) {
    return { age: years, period: years === 1 ? 'year' : 'years' };
  }

  let months =
    end.getMonth() -
    start.getMonth() +
    12 * (end.getFullYear() - start.getFullYear());
  if (months >= 1) {
    return { age: months, period: months === 1 ? 'month' : 'months' };
  }

  let days = end.getDate() - start.getDate();
  return { age: days, period: days === 1 ? 'day' : 'days' };
};

const getExpectedDeathDate = (corpse: Corpse): string =>
  corpse.deathDate.toLocaleDateString(undefined, {
    month: 'long',
    year: 'numeric',
  });

const getFullName = (corpse: Corpse): string =>
  corpse.qualifier ? `${corpse.name} (${corpse.qualifier})` : corpse.name;

const getLifeDates = (corpse: Corpse): string =>
  corpse.birthDate
    ? `${corpse.birthDate.getFullYear()} - ${corpse.deathDate.getFullYear()}`
    : `${corpse.deathDate.getFullYear()}`;

const getObituary = (corpse: Corpse, today: Date): string => {
  let obituary = '';

  const dead = isDead(corpse, today);
  if (dead) {
    const { age, period } = getAge(corpse.deathDate, today);
    const message = age === 0 ? 'today' : `${age} ${period} ago`;
    obituary += `Killed by Microsoft ${message}, `;
  } else {
    const { age, period } = getAge(today, corpse.deathDate);
    obituary += `To be killed by Microsoft in ${age} ${period}, `;
  }

  obituary += `${corpse.name} ${dead ? 'was' : 'is'} ${corpse.description}.`;

  if (dead && corpse.birthDate) {
    const { age, period } = getAge(corpse.birthDate, corpse.deathDate);
    obituary += ` It was ${age} ${period} old.`;
  }

  return obituary;
};

const isDead = (corpse: Corpse, today: Date): boolean =>
  corpse.deathDate <= today;

export {
  Corpse,
  getExpectedDeathDate,
  getFullName,
  getLifeDates,
  getObituary,
  isDead,
};

File: src/App/Models/Corpse.cs
namespace VictorFrye.Graveyard.Models;

using System.Text;

using Microsoft.FluentUI.AspNetCore.Components;

public class Corpse(string name, string? qualifier, DateOnly? birthDate, DateOnly deathDate, string description, string link)
{
    public string Name { get; init; } = name;
    public string? Qualifier { get; init; } = qualifier;
    public DateOnly? BirthDate { get; init; } = birthDate;
    public DateOnly DeathDate { get; init; } = deathDate;
    public string Description { get; init; } = description;
    public string Link { get; init; } = link;

    public bool IsDead(DateOnly today) => DeathDate <= today;

    public string GetExpectedDeathDate() => $"{DeathDate:MMMM yyyy}";

    public string GetLifeDates() => BirthDate is null ? $"{DeathDate:yyyy}" : $"{BirthDate.Value:yyyy} - {DeathDate:yyyy}";

    public string GetFullName() => Qualifier is null ? Name : $"{Name} ({Qualifier})";

    public string GetObituary(DateOnly today)
    {
        var obituary = new StringBuilder();
        var isDead = IsDead(today);

        if (isDead)
        {
            var (age, period) = GetAge(DeathDate, today);
            var message = age == 0 ? "today" : $"{age} {period} ago";
            obituary.Append($"Killed by Microsoft {message}, ");
        }
        else
        {
            var (age, period) = GetAge(today, DeathDate);
            obituary.Append($"To be killed by Microsoft in {age} {period}, ");
        }

        obituary.Append($"{Name} {(isDead ? "was" : "is")} {Description}.");

        if (isDead && BirthDate is not null)
        {
            var (age, period) = GetAge(BirthDate.Value, DeathDate);
            obituary.Append($" It was {age} {period} old.");
        }

        return obituary.ToString();
    }

    internal static (int Age, string Period) GetAge(DateOnly start, DateOnly end)
    {
        var years = end.Year - start.Year;

        if (end.Month < start.Month || (end.Month == start.Month && end.Day < start.Day))
        {
            years--;
        }

        if (years >= 1)
        {
            return (years, years == 1 ? "year" : "years");
        }

        var months = end.Month - start.Month + (12 * (end.Year - start.Year));

        if (months >= 1)
        {
            return (months, months == 1 ? "month" : "months");
        }

        var days = end.Day - start.Day;

        return (days, days == 1 ? "day" : "days");
    }
}
